pkg/deploy/envParams.go

```diff
@@ -75,22 +85,28 @@ func ApplyParams(componentPath string, imageParamsMap map[string]string, extraPa
 		return err
 	}
 
+	count := 0
+
 	// 1. Update images with env variables
 	// e.g "odh-kuberay-operator-controller-image": "RELATED_IMAGE_ODH_KUBERAY_OPERATOR_CONTROLLER_IMAGE",
 	for i := range paramsEnvMap {
 		relatedImageValue := os.Getenv(imageParamsMap[i])
 		if relatedImageValue != "" {
-			paramsEnvMap[i] = relatedImageValue
+			count += updateMap(&paramsEnvMap, i, relatedImageValue)
 		}
 	}
 
 	// 2. Update other fileds with extraParamsMap which are not carried from component
 	for _, extraParamsMap := range extraParamsMaps {
 		for eKey, eValue := range extraParamsMap {
-			paramsEnvMap[eKey] = eValue
+			count += updateMap(&paramsEnvMap, eKey, eValue)
 		}
 	}
 
+	if count == 0 {
```

Comment:
need a count?
or just a true/false enough?
if !needUpdateImage && !needUpdateEnv {
return nil
}

Reply:
I do not a way to accumulate booleans. With every individual check it looks pretty clumsy. But actually I missed for some reason that bitwise operations work in go, so will replace += with |= to be safe from overflow :) (practically impossible here)

Reply:
but to call writeParamsToTmp() or not, this is based on if any key-value  from passing-in map does not match the key-value from params.env, isnt it? regardless all of them mismatch or just one of them mismatch.

as for writeParamsToTmp(), it is a full dump all key-values into the temp. file

so i still do not get the reason behind why need count them.

if you have a first mistmatch in the updateMap, mark needUpdateImage/needUpdateEnv to true, isnt that enough? something like 
```
func updateMap(m *map[string]string, key, val string) int {
	old := (*m)[key]
	if old != val {
		(*m)[key] = val
        needUpdateImage = true
	}
}
```

Reply:
It should be a package scope variable then, which is not nice as well. But if you prefer that, I'll  change that way

Reply:
And negated condition :)

Reply:
> so i still do not get the reason behind why need count them.

There is no need to really count, it acts as boolean actually just to accumulate results (a bit like multierror does for error) since I cannot do var += true and var += false which is true if one of that was true

Reply:
> It should be a package scope variable then, which is not nice as well. But if you prefer that, I'll change that way

Actually, it's more serious issue. Since it addresses the possible race, package scoped variable requires synchronized access then (mutex). Basically, setting it "true" from two invocations does not make real harm with the implementation, since tmpfile name is unique, update of boolean is atomic and params.env file update with os.Rename() is atomic as well, but it's not nice too.

Reply:
> Actually, it's more serious issue. Since it addresses the possible race, package scoped variable requires ~~synchronized access then (mutex). Basically, setting it "true" from two invocations does not make real harm with the implementation, since tmpfile name is unique, update of boolean is atomic and params.env file update with os.Rename() is atomic as well, but it's not nice too.~~

Forget what I said. It's an issue since there are 2 places of writes: before the loop and when a field gets updated. They may interlace and previously set "true" be reset with next invocation "false".

Reply:
In general, these params.env corrections look like one time initialization operation and should not be called for every reconciliation.

Reply:
> In general, these params.env corrections look like one time initialization operation and should not be called for every reconciliation.

https://github.com/opendatahub-io/opendatahub-operator/pull/1191 like that, what do you think?